src/session_manager.rs

```diff
@@ -0,0 +1,1055 @@
+                        other => {
+                            self_.sessions_handled.set(self_.sessions_handled.get() + 1);
+
+                            if self_.initial_sessions_to_handle.get() == 1 {
+                                if APPLICATION_OPTS.get().unwrap().test_dc
+                                    != database_info.use_test_dc
+                                {
+                                    self.tear_down_client(client_id, || {}, || {});
+                                    self.add_new_session(!database_info.use_test_dc);
+                                } else {
+                                    // Overwrite ClientInfo.
+                                    self_.clients.borrow_mut().insert(
+                                        client_id,
+                                        ClientInfo::Auth {
+                                            maybe_authorized: false,
+                                            is_active,
+                                            database_info: database_info.clone(),
+                                        },
+                                    );
+
+                                    self_.login.login_client(client_id, database_info);
+
+                                    self_
+                                        .login
+                                        .set_authorization_state(update.authorization_state);
+                                }
+                            } else {
+                                match other {
+                                    AuthorizationState::LoggingOut => {
+                                        log::warn!(
+                                            "Unexpected authorization state '{:?}' for previously \
+                                            logged in session (db directory = {}). This is \
+                                            probably because you killed the session on another \
+                                            device.",
+                                            other,
+                                            database_info.directory_base_name,
+                                        );
+                                        // TODO: Show a toast notification.
+
+                                        // Just remove the database. No need to logout explicitly. This
+                                        // is already ongoing.
+                                        RUNTIME.spawn(async move {
+                                            delete_database_dir(database_info.directory_base_name)
+                                                .await
+                                                .unwrap();
+                                        });
+                                    }
+                                    _ => {
+                                        log::warn!(
+                                            "Unexpected authorization state '{:?}' for previously
```

Comment:
Is it really unexpected? If there's even an authorization state just for that I think it's indeed expected.